Lib/fontPens/penTools.py

```diff
@@ -55,24 +71,144 @@ def getCubicPoint(t, pt0, pt1, pt2, pt3):
+    """
+    (x0, y0), (x1, y1) = pt0, pt1
+    cx = (x1 - x0) * 3
+    cy = (y1 - y0) * 3
+    bx = (pt2[0] - x1) * 3 - cx
+    by = (pt2[1] - y1) * 3 - cy
+    ax = pt3[0] - x0 - cx - bx
+    ay = pt3[1] - y0 - cy - by
+    path = []
+    for t in ts:
+        t3 = t ** 3
+        t2 = t * t
+        x = ax * t3 + bx * t2 + cx * t + x0
+        y = ay * t3 + by * t2 + cy * t + y0
+        path.append((x, y))
+    return path
+
+
 def estimateCubicCurveLength(pt0, pt1, pt2, pt3, precision=10):
     """
     Estimate the length of this curve by iterating
     through it and averaging the length of the flat bits.
+
+    >>> estimateCubicCurveLength((0, 0), (0, 0), (0, 0), (0, 0), 1000)
+    0.0
+    >>> estimateCubicCurveLength((0, 0), (0, 0), (120, 0), (120, 0), 1000)
+    120.0
+    >>> estimateCubicCurveLength((0, 0), (50, 0), (80, 0), (120, 0), 1000)
+    120.0
+    >>> estimateCubicCurveLength((0, 0), (50, -10), (80, 50), (120, 40), 1)
+    126.49110640673517
+    >>> estimateCubicCurveLength((0, 0), (50, -10), (80, 50), (120, 40), 1000)
+    130.4488917899906
+    """
+    
+    length = 0
+    step = 1.0 / precision
+    points = getCubicPoints([f * step for f in range(precision + 1)], pt0, pt1, pt2, pt3)
+    for i in range(len(points) - 1):
+        pta = points[i]
+        ptb = points[i + 1]
+        length += distance(pta, ptb)
+    return length
+
+
+def estimateQuadraticCurveLength(pt0, pt1, pt2, precision=10):
+    """
+    Estimate the length of this curve by iterating
+    through it and averaging the length of the flat bits.
+
+    >>> estimateQuadraticCurveLength((0, 0), (0, 0), (0, 0)) # empty segment
+    0.0
+    >>> estimateQuadraticCurveLength((0, 0), (50, 0), (80, 0)) # collinear points
+    80.0
+    >>> estimateQuadraticCurveLength((0, 0), (50, 20), (100, 40)) # collinear points
+    107.70329614269008
+    >>> estimateQuadraticCurveLength((0, 0), (0, 100), (100, 0))
+    153.6861437729263
+    >>> estimateQuadraticCurveLength((0, 0), (50, -10), (80, 50))
+    102.4601733446439
+    >>> estimateQuadraticCurveLength((0, 0), (40, 0), (-40, 0)) # collinear points, control point outside
+    66.39999999999999
+    >>> estimateQuadraticCurveLength((0, 0), (40, 0), (0, 0)) # collinear points, looping back
+    40.0
     """
     points = []
     length = 0
     step = 1.0 / precision
     factors = range(0, precision + 1)
     for i in factors:
-        points.append(getCubicPoint(i * step, pt0, pt1, pt2, pt3))
+        points.append(getQuadraticPoint(i * step, pt0, pt1, pt2))
     for i in range(len(points) - 1):
         pta = points[i]
         ptb = points[i + 1]
         length += distance(pta, ptb)
     return length
 
 
+def getQuadraticCurveLength(pt0, pt1, pt2, approximate_fallback=False):
```

Comment:
and why do you need this `getQuadraticCurveLength` wrapper?

Reply:
I thought to keep the different functions there in that file, but it's not really needed. But then the `estimateCubicCurveLength` function probably should also be replaced? Didn't want to change Just's existing code too much ... the quadratic function is new so nobody is using it yet.